src/RawPOSIX/src/safeposix/dispatcher.rs

```diff
@@ -250,7 +256,10 @@ pub fn lind_syscall_api(
 
         READ_SYSCALL => {
             let fd = arg1 as i32;
-            let buf = (start_address + arg2) as *mut u8;
+            let buf = match check_and_convert_addr(cageid, start_address + arg2) {
+                Ok(addr) => addr as *mut u8,
+                Err(errno) => return syscall_error(errno, "read", "invalid buffer address"),
+            };
             let count = arg3 as usize;
 
             interface::cagetable_getref(cageid)
```

Comment:
like here we have to get it again, so if you get it before your function and save it as variable, ie "cage" then you can pass it into your function i believe.